core/pkg/artifacts/saver.go

```diff
@@ -512,34 +513,129 @@ func multiPartRequest(path string) ([]gql.UploadPartsInput, error) {
-		if err != nil && err != io.EOF {
-			return nil, err
-		}
-		if bytesRead == 0 {
-			break
+func computeMultipartHashes(logger *observability.CoreLogger, path string, chunkSize int64, numWorkers int) ([]gql.UploadPartsInput, error) {
+	if numWorkers < 1 {
+		return nil, fmt.Errorf("number of workers is less than 1: %d", numWorkers)
+	}
+	if chunkSize < 1 {
+		return nil, fmt.Errorf("chunk size is less than 1: %d", chunkSize)
+	}
+
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file size for path %s: %w", path, err)
+	}
+	fileSize := fileInfo.Size()
+	if fileSize == 0 {
+		return nil, fmt.Errorf("file size is 0") // should never happen when the multipart threshold check is used properly
+	}
+	if fileSize < chunkSize {
+		return nil, fmt.Errorf("file size is less than chunk size: %d < %d", fileSize, chunkSize)
+	}
+
+	numParts := int(fileSize / chunkSize)
+	if fileSize%chunkSize != 0 {
+		numParts++
+	}
+	numWorkers = min(numWorkers, numParts)
+
+	partsPerWorker := numParts / numWorkers
+	remainingParts := numParts % numWorkers
+
+	workerTasks := make([]struct {
+		startPart int // inclusive
+		endPart   int // exclusive
+	}, numWorkers)
+	// NOTE: We start from 0, and only switch to 1-indexed when assiging the partNumber for server request
+	startPart := 0
+	for i := 0; i < numWorkers; i++ {
+		endPart := startPart + partsPerWorker
+		// Split the remaining parts evenly among the workers instead of doubling the work for the last worker.
+		if remainingParts > 0 {
+			endPart++
+			remainingParts--
 		}
-		partsInfo = append(partsInfo, gql.UploadPartsInput{
-			PartNumber: partNumber,
-			HexMD5:     hashencode.ComputeHexMD5(buffer[:bytesRead]),
-		})
-		partNumber++
+		workerTasks[i].startPart = startPart
+		workerTasks[i].endPart = endPart
+		startPart = endPart
+	}
+
+	partsInfo := make([]gql.UploadPartsInput, numParts)
+
+	startTime := time.Now()
+	var wg sync.WaitGroup
+	errChan := make(chan error, numWorkers)
+	wg.Add(numWorkers)
+	for worker, workerTask := range workerTasks {
+		go func(worker, startPart, endPart int) {
+			defer wg.Done()
+			hashFileParts(worker, startPart, endPart, path, chunkSize, fileSize, partsInfo, errChan)
+		}(worker, workerTask.startPart, workerTask.endPart)
 	}
+
+	wg.Wait()
+	close(errChan)
+	for err := range errChan {
+		return nil, err
+	}
+
+	hashTime := time.Since(startTime)
+	hashSpeedMBps := float64(fileSize) / (1024 * 1024) / hashTime.Seconds()
+	logger.Debug("Computed multipart hashes",
+		"hashTimeMs", hashTime.Milliseconds(),
+		"hashSpeedMBps", hashSpeedMBps,
+		"numWorkers", numWorkers,
+		"numParts", len(partsInfo),
+		"fileSize", fileSize,
+		"chunkSize", chunkSize,
+	)
 	return partsInfo, nil
 }
 
+// hashFileParts hash the md5 of consecutive parts of the file in serial.
+func hashFileParts(worker, startPart, endPart int, path string, chunkSize, fileSize int64, partsInfo []gql.UploadPartsInput, errChan chan<- error) {
```

Comment:
nit: in Go you have to limit line length manually, the formatter is very lax. Please put the parameter sets on their own lines (e.g. `worker, startPart, endPart int` would be one line).

This is also quite a lot of parameters. Consider creating a helper `multipartHashWorker` struct and making this one of its methods.